launch/launch/substitutions/launch_configuration.py

```diff
@@ -0,0 +1,87 @@
+# Copyright 2018 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Module for the LaunchConfiguration substitution."""
+
+import collections
+from typing import Any
+from typing import Iterable
+from typing import List
+from typing import Optional
+from typing import Text
+from typing import Union
+
+from .substitution_failure import SubstitutionFailure
+from ..launch_context import LaunchContext
+from ..some_substitutions_type import SomeSubstitutionsType
+from ..substitution import Substitution
+from ..utilities import ensure_argument_type
+
+
+class LaunchConfiguration(Substitution):
+    """Substitution that can access launch configuration variables."""
+
+    def __init__(
+        self,
+        variable_name: SomeSubstitutionsType,
+        *,
+        default: Optional[Union[Any, Iterable[Any]]] = None,
+    ) -> None:
+        """Constructor."""
+        super().__init__()
+
+        ensure_argument_type(variable_name, str, 'variable_name', 'LaunchConfiguration')
+
+        from ..utilities import normalize_to_list_of_substitutions
+        self.__variable_name = normalize_to_list_of_substitutions(variable_name)
+        if default is None:
+            self.__default = default
+        else:
+            # convert any items in default that are not a Substitution or str to a str
```

Comment:
I want to note that it might trip users up that the default gets converted to a string underneath. It might be better to just enforce it's passed in as a string (if it's not a substitution) rather than doing it ourselves, so that a user that passes in a dictionary or something doesn't expect to be able to retrieve a dictionary later.

Reply:
It's a trade-off, as it's convenient to pass literals or the result of expressions to the `SetLaunchConfiguration` action directly. I can change it.

Reply:
I dunno, it's pretty convenient. I think I'd prefer to leave it.

If you want, I'll change it anyways, but what if I just improved the documentation a bit to be clear that the default must be a string in the end, so non-text/non-substitution things are converted to text implicitly?

Reply:
yeah I already clarified a bit in https://github.com/ros2/launch/pull/74/commits/760fe46c6c886f9446e7c84bd726258d1081be83#diff-55c61397a0c771113a1e78e3143e31fdR77. I can see the convenience, so let's leave it; maybe one day we will store dicts as dicts even.